syft/serde/serde.py

```diff
@@ -0,0 +1,476 @@
+
+    if len(compress_stream) < len(decompressed_input_bin):
+        return compress_scheme.to_bytes(1, byteorder="big") + compress_stream
+    else:
+        return NO_COMPRESSION.to_bytes(1, byteorder="big") + decompressed_input_bin
+
+
+def _decompress(binary: bin) -> bin:
+    """
+    This function decompresses a binary using the scheme defined in the first byte of the input
+
+    Args:
+        binary (bin): a compressed binary
+
+    Returns:
+        bin: decompressed binary
+
+    """
+
+    # check the 1-byte header to check the compression scheme used
+    compress_scheme = binary[0]
+
+    # remove the 1-byte header from the input stream
+    binary = binary[1:]
+    # 1)  Decompress or return the original stream
+    if compress_scheme == LZ4:
+        return lz4.frame.decompress(binary)
+    elif compress_scheme == ZSTD:
+        return zstd.decompress(binary)
+    elif compress_scheme == NO_COMPRESSION:
+        return binary
+    else:
+        raise CompressionNotFoundException(
+            "compression scheme not found for" " compression code:" + str(compress_scheme)
+        )
+
+
+# High Level Simplification Router
+
+
+def _simplify(obj: object) -> object:
+    """
+    This function takes an object as input and returns a simple
+    Python object which is supported by the chosen serialization
+    method (such as JSON or msgpack). The reason we have this function
+    is that some objects are either NOT supported by high level (fast)
+    serializers OR the high level serializers don't support the fastest
+    form of serialization. For example, PyTorch tensors have custom pickle
+    functionality thus its better to pre-serialize PyTorch tensors using
+    pickle and then serialize the binary in with the rest of the message
+    being sent.
+
+    Args:
+        obj: an object which may need to be simplified
+
+    Returns:
+        obj: an simple Python object which msgpack can serialize
+
+    Raises:
+        ValueError: if `move_this` or `in_front_of_that` are not both single ASCII
+        characters.
+
+    """
+
+    try:
+        # check to see if there is a simplifier
+        # for this type. If there is, run return
+        # the simplified object
+        current_type = type(obj)
+
+        result = (simplifiers[current_type][0], simplifiers[current_type][1](obj))
+
+        return result
+
+    except KeyError:
+
+        # if there is not a simplifier for this
+        # object, then the object is already a
+        # simple python object and we can just
+        # return it
+        return obj
+
+
+def _force_full_simplify(obj: object) -> object:
+    current_type = type(obj)
+
+    if current_type in forced_full_simplifiers:
+
+        left = forced_full_simplifiers[current_type][0]
+
+        right = forced_full_simplifiers[current_type][1]
+
+        right = right(obj)
+
+        result = (left, right)
+    else:
+        result = _simplify(obj)
+
+    return result
+
+
+simplifiers = {
```

Comment:
I'm not sure if @robert-wagner @LaRiffle agree, but I would love to get rid of this dict, or maybe store simplifiers and detailers in the same dict.

Reply:
Do we need to keep this for efficiency reasons?

Reply:
given that we really need it, my vote is: make a list of pairs containing (simplifier, detailer) pairs, a dict can later be built automatically using this list during init

Reply:
I hate this dict but it does have some advantages for performance. That being said, if we could get rid of it I would be very happy

Reply:
I would suggest doing something like in hook_args: the dict is built dynamically during execution: so each time you add an entry you have a tiny overhead and then you have the dict look-up speed. This would use a list of pairs as suggested by @mari-linhares to build on this

Reply:
sgtm

Reply:
Hi @mari-linhares and @LaRiffle, 

I am trying to implement the hook_args @LaRiffle suggested. A bit lost to where to start.
I checked the file frameworks/torch/hook/hook_args.py.
Is it similar?
Still confused, would love to get some guidance on it? Thanks
Is the args would be same at each execution?

Reply:
So the mechanism would be the following:
Instead of 
```
simplifiers = {
    torch.Tensor: [0, _simplify_torch_tensor],
    torch.nn.Parameter: [1, _simplify_torch_parameter],
    tuple: [2, _simplify_collection],
    list: [3, _simplify_collection],
```
Have
```
simplifiers = {}
map_simplifiers = {
    torch.Tensor: (_simplify_torch_tensor, _detail_torch_tensor),
    torch.nn.Parameter: (_simplify_torch_parameter, _detail_torch_parameter),
    etc
detailers = []
```
Then in def _simplify, check if simplifiers[type] succeeds, in case of key Error,  create
simplifiers[type]= [index, func] where func = map_simplifiers[key][0] and append map_simplifiers[key][1] in a detailer list ; also increment index

This is the first step, does it make sense?

Reply:
I think we can do this change in a new PR :relaxed:

Reply:
Ok I ll open another one for that right after closing this one I guess.
Thanks @LaRiffle , I do get it and working on it.